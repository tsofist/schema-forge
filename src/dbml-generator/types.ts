import type { ArrayMay, HexString } from '@tsofist/stem';
import type { SchemaDefinitionInfoForType } from '../definition-info/types';

/**
 * DBML Project Source
 */
export interface DBMLProjectScope {
    schemaId: string;
    scopeName?: string;
    definitions: SchemaDefinitionInfoForType[];
    comment?: string;
}

/**
 * DBML Project Metadata
 */
export interface DBMLProjectMeta {
    name: string;
    note?: string;
    comment?: string;
}

/**
 * DBML Generator Options
 */
export interface DBMLGeneratorOptions {
    meta?: DBMLProjectMeta;
    // /**
    //  * @deprecated use notes
    //  */
    includeNotes?: boolean;
    columnsOrder?: string[];
    // todo implement
    // notes?:
    //     | boolean
    //     | {
    //           project?: boolean;
    //           tables?: boolean;
    //           columns?: boolean;
    //           indexes?: boolean;
    //       };
    // enforceSnakeCase?: boolean; // todo implement
    // columnTypeInference?: {
    //     arraysAsJSONB?: boolean; // todo implement
    // };
}

/**
 * Database index type list.
 *
 * @see https://www.postgresql.org/docs/current/indexes-types.html PostgreSQL
 */
export const DBMLIndexTypeList = [
    //
    'btree',
    'hash',
    'gin',
    'gist',
    'spgist',
    'brin',
] as const;

/**
 * Database column type list.
 */
export const DBMLColumnTypeList = [
    'uuid',
    'integer',
    'bigint',
    'smallint',
    'numeric',
    'numeric(12,2)',
    'numeric(15,2)',
    'numeric(30,18)',
    'float',
    'float8',
    'boolean',
    'text',
    'date',
    'time',
    'timestamp',
    'timestamptz',
    //
    'uuid[]',
    'integer[]',
    'bigint[]',
    'smallint[]',
    'numeric[]',
    'numeric(12,2)[]',
    'numeric(15,2)[]',
    'numeric(30,18)[]',
    'float[]',
    'float8[]',
    'boolean[]',
    'text[]',
    'date[]',
    'time[]',
    'timestamp[]',
    'timestamptz[]',
] as const;

/**
 * Table column type.
 */
export type DBMLColumnType = (typeof DBMLColumnTypeList)[number];

/**
 * Database index type.
 *
 * @see https://www.postgresql.org/docs/current/indexes-types.html PostgreSQL
 * @default 'btree'
 */
export type DBMLIndexType = (typeof DBMLIndexTypeList)[number];

/**
 * Database index options.
 *
 * @see https://dbml.dbdiagram.io/docs/#index-definition dbml spec
 */
export type DBMLIndexOptions = {
    /**
     * Type of index.
     * @default 'btree'
     */
    type?: DBMLIndexType;
    /**
     * Name of index.
     * @default 'ix_[schema]_[table]_[column]'
     */
    name?: string;
    /**
     * If true, the index will be unique.
     * @default false
     */
    unique?: boolean;
    /**
     * Is it a primary key
     * @default false
     */
    pk?: boolean;
    /**
     * Any important notes.
     * Can be used in DDL.
     */
    note?: string;
    /**
     * Any important developer comments.
     * Can't be used in DDL.
     */
    comment?: string;
};

export type DBMLIndexOptionsDef<B extends boolean = boolean> = ArrayMay<
    DBMLIndexOptions | string | B
>;

export type DBMLForeignKeyOptionsDef = boolean;
export type DBMLForeignKeyOptions = DBMLForeignKeyOptionsDef;

/**
 * Database entity options.
 *
 * @see https://dbml.dbdiagram.io/docs/#table-definition dbml spec
 */
export type DBMLEntityOptions = {
    /**
     * Table name.
     *
     * @default '[schema].[table]'
     */
    name?: string;
    /**
     * Indexes of the table.
     * They can be overridden/added at the column level.
     *
     * Use `true` to apply default index options.
     * Use string literal to set index name.
     */
    indexes?: {
        [field: string]: DBMLIndexOptionsDef<true>;
    };
    /**
     * Any important notes.
     * Can be used in DDL.
     */
    note?: string;
    /**
     * Any important developer comments.
     * Can't be used in DDL.
     */
    comment?: string;
    /**
     * Table name alias.
     */
    alias?: string;
    /**
     * Settings.
     */
    settings?: { headercolor: HexString };
};

export type DBMLEntityOptionsDef = DBMLEntityOptions | string;

export type DBMLColumnOptions = {
    /**
     * Is it a primary key
     *
     * @default false
     */
    pk?: boolean;
    /**
     * Column type.
     *
     * If `undefined`, the type will be inferred automatically.
     */
    type?: DBMLColumnType;
};

export type DBMLEnumOptionsDef = boolean | string;
