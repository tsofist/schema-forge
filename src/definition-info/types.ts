import type { SchemaForgeDefinitionRef } from '../types';

/**
 * Kind of schema definition information
 */
export enum SchemaDefinitionInfoKind {
    Type,
    API,
    APIMethodResult,
    APIMethodArguments,
}

/**
 * Schema definition information
 */
export type SchemaDefinitionInfo =
    | SchemaDefinitionInfoForType
    | SchemaDefinitionInfoForAPIInterface
    | SchemaDefinitionInfoForAPIMethodArguments
    | SchemaDefinitionInfoForAPIMethodResult;

export interface SchemaDefinitionInfoForType extends SDIBase {
    kind: SchemaDefinitionInfoKind.Type;
    type: string;
}

export interface SchemaDefinitionInfoForAPIInterface extends SDIBase {
    kind: SchemaDefinitionInfoKind.API;
    interface: string;
}

export interface SchemaDefinitionInfoForAPIMethodArguments extends SDIBase {
    kind: SchemaDefinitionInfoKind.APIMethodArguments;
    interface: string;
    method: string;
}

export interface SchemaDefinitionInfoForAPIMethodResult extends SDIBase {
    kind: SchemaDefinitionInfoKind.APIMethodResult;
    interface: string;
    method: string;
}

interface SDIBase {
    name: string;
    schemaId: string;
    ref: SchemaForgeDefinitionRef;
    kind: SchemaDefinitionInfoKind;
}

/** @deprecated */
export const _SDS_SUFFIX_API = '_InterfaceDeclaration';
/** @deprecated */
export const _SDS_SUFFIX_METHOD_ARGS = '_Args';
/** @deprecated */
export const _SDS_SUFFIX_METHOD_RES = '_Result';

export const SDS_SUFFIX_API = '__APIInterface';
export const SDS_SUFFIX_MEMBER = '__APIMember';
export const SDS_SUFFIX_METHOD_ARGS = '__APIMethodArgs';
export const SDS_SUFFIX_METHOD_RES = '__APIMethodResult';

export type APIInterfaceSDS = `${string}${typeof SDS_SUFFIX_API}`;
export type APIMemberSDS = `${string}_${string}${typeof SDS_SUFFIX_MEMBER}`;
export type APIMethodArgsSDS = `${string}_${string}${typeof SDS_SUFFIX_METHOD_ARGS}`;
export type APIMethodResultSDS = `${string}_${string}${typeof SDS_SUFFIX_METHOD_RES}`;
