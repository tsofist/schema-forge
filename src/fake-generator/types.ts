import * as fakerModule from '@faker-js/faker';
import type { ArrayMay } from '@tsofist/stem';
import { JSONSchemaFaker, type JSONSchemaFakerOptions } from 'json-schema-faker';
import type { SchemaForgeRegistry } from '../schema-registry/types';

export type SetupFakerModules = (faker: fakerModule.Faker) => object;
export type FakerRangeNum = Parameters<fakerModule.HelpersModule['rangeToNumber']>[0];

export type FakeGeneratorLocaleName = keyof typeof fakerModule.allLocales;
export interface FakeGeneratorOptions extends JSONSchemaFakerOptions {
    locale?: ArrayMay<FakeGeneratorLocaleName>;
    setupFakerModules?: SetupFakerModules[];
}

export interface FakeGeneratorHost {
    readonly schemaRegistry: SchemaForgeRegistry;
    readonly faker: fakerModule.Faker;
    readonly generator: typeof JSONSchemaFaker;
    readonly rebuild: () => this;
}
