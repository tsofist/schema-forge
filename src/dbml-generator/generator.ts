import type { PRec } from '@tsofist/stem';
import { asArray } from '@tsofist/stem/lib/as-array';
import { isEqualKeys } from '@tsofist/stem/lib/equal-keys';
import { raise } from '@tsofist/stem/lib/error';
import { entries } from '@tsofist/stem/lib/object/entries';
import { snakeCase } from '@tsofist/stem/lib/string/case/snake';
import { compareStringsAsc } from '@tsofist/stem/lib/string/compare';
import { substr } from '@tsofist/stem/lib/string/substr';
import { TextBuilder } from '@tsofist/stem/lib/string/text-builder';
import type { JSONSchema7 as Schema } from 'json-schema';
import type { SchemaDefinitionInfoForType } from '../definition-info/types';
import {
    createSchemaDereferenceSharedCache,
    SchemaDereferenceSharedCache,
} from '../schema-dereference/cache';
import { dereferenceSchema } from '../schema-dereference/dereference';
import type { SchemaForgeRegistry } from '../schema-registry/types';
import type { ForgedEntitySchema, ForgedPropertySchema } from '../types';
import type {
    DBMLEntityOptions,
    DBMLGeneratorOptions,
    DBMLIndexOptions,
    DBMLIndexOptionsDef,
    DBMLProjectScope,
} from './types';

export function generateDBMLSpec(
    schemaRegistry: SchemaForgeRegistry,
    scopes: DBMLProjectScope[],
    options: DBMLGeneratorOptions = {},
) {
    const text = new TextBuilder();
    const tables = new Map<
        string, // table name
        TableSpec
    >();
    const groups = new Map<
        string, // group name
        string[] // table names
    >();
    const sources = new Map<
        string, // scope name
        DBMLProjectScope
    >();
    const dereferencedRootSchemas: Map<string, Schema> = new Map();

    if (options.meta?.comment) {
        text.push(`// ${options.meta?.comment}`);
    }
    text.push(`Project ${options.meta?.name ?? 'Scratch'} {`);
    text.push(`database_type: 'PostgreSQL'`, 1);
    if (options.meta?.note) {
        text.push('');
        text.push(buildNote(options.meta.note), 1);
    }
    text.push('}');
    text.push(``);

    for (const scope of scopes.sort((a, b) => {
        return compareStringsAsc(a.scopeName ?? DefaultScopeName, b.scopeName ?? DefaultScopeName);
    })) {
        if (sources.has(scope.scopeName ?? DefaultScopeName)) {
            raise(`Duplicate scope: ${scope.scopeName ?? '[ unnamed ]'}`);
        } else {
            sources.set(scope.scopeName ?? DefaultScopeName, scope);
        }
    }

    for (const { definitions, scopeName = DefaultScopeName, schemaId } of sources.values()) {
        let dereferencedRootSchema: Schema = dereferencedRootSchemas.get(schemaId)!;
        if (!dereferencedRootSchemas.has(schemaId)) {
            const root =
                schemaRegistry.getRootSchema(schemaId) ||
                raise(`Root schema ${schemaId} not found`);
            dereferencedRootSchema =
                dereferenceSchema(root, {
                    throwOnDereferenceFailure: true,
                    sharedCacheStorage:
                        DereferenceSchemaCache ||
                        (DereferenceSchemaCache = createSchemaDereferenceSharedCache()),
                }) || raise(`Failed to dereference root schema for ${schemaId}`);
            dereferencedRootSchemas.set(schemaId, dereferencedRootSchema);
        }

        for (const definition of definitions) {
            const tableSpec = generateTable(definition, options, dereferencedRootSchema);
            if (tableSpec) {
                if (tables.has(tableSpec.name)) {
                    raise(`Definitions contain duplicate table name: ${tableSpec.name}`);
                }

                tables.set(tableSpec.name, tableSpec);

                let groupTables = groups.get(scopeName);
                if (!groupTables) {
                    groupTables = [];
                    groups.set(scopeName, groupTables);
                }

                groupTables.push(tableSpec.name);
            }
        }
    }

    const defaultGroupOnly = groups.size === 1 && groups.has(DefaultScopeName);

    for (const [groupName, groupTableNames] of groups.entries()) {
        for (const tableName of groupTableNames.sort(compareStringsAsc)) {
            const table = tables.get(tableName)!;
            text.push(table.value);
            text.push(``);
        }

        if (!defaultGroupOnly) {
            const scope = sources.get(groupName);
            if (scope) {
                const comment = scope.comment;
                if (comment) {
                    text.push(`// ${comment}`);
                }
                text.push(`TableGroup ${groupName} {`);
                for (const tableName of groupTableNames.sort(compareStringsAsc)) {
                    text.push(tableName, 1);
                }
                text.push(`}`);
                text.push(``);
            }
        }
    }

    return text.stringify();
}

function generateTable(
    info: SchemaDefinitionInfoForType,
    options: DBMLGeneratorOptions,
    dereferencedRootSchema: Schema,
): TableSpec | undefined {
    const includeNotes = options.includeNotes ?? false;
    const entityTypeName = info.type;
    const columnsOrder = options.columnsOrder || DefaultColumnsOrder;
    const entitySchema = (dereferencedRootSchema.definitions || dereferencedRootSchema.$defs)?.[
        entityTypeName
    ];
    if (!entitySchema) return;
    if (typeof entitySchema !== 'object')
        raise(`Invalid definition type for table: ${typeof entitySchema} (expected object)`);

    const tableName = readDBEntityName(entitySchema);
    if (!tableName) return;

    const { properties, required } = listProperties(entitySchema);
    const columns = generateColumns(
        properties,
        required,
        includeNotes,
        info.schemaId,
        tableName,
        columnsOrder,
    );
    const indexes = generateIndexes(
        tableName,
        properties,
        readDBEntityIndexes(entitySchema),
        includeNotes,
        info.schemaId,
    );

    const text = new TextBuilder();

    if (entitySchema.$comment) {
        text.push(stringifyComment(entitySchema.$comment));
    }

    text.push(`Table ${tableName} {`);
    text.push(columns, 1);

    if (indexes.size) {
        text.push('');
        text.push('indexes {', 1);
        text.push(indexes, 2);
        text.push('}', 1);
    }

    if (entitySchema.description) {
        text.push('');
        text.push(buildNote(entitySchema.description), 1);
    }

    text.push('}');

    return {
        name: tableName,
        value: text.stringify(),
    };
}

function generateColumns(
    properties: PRec<ForgedPropertySchema>,
    requiredFields: string[],
    notes: boolean,
    schemaId: string,
    tableName: string,
    columnsOrder: string[],
): TextBuilder {
    const text = new TextBuilder();
    let count = 0;

    for (const [key, property] of entries(properties).sort(([keyA], [keyB]) => {
        const indexA = columnsOrder.indexOf(keyA);
        const indexB = columnsOrder.indexOf(keyB);
        if (indexA === -1 && indexB === -1) return 0;
        else if (indexA === -1) return 1;
        else if (indexB === -1) return -1;

        return indexA - indexB;
    })) {
        if (property === undefined) continue;

        const isRequired = requiredFields.includes(key);
        const isColumnNullable = isNullable(property);
        const columnType = getDBType(property, schemaId);
        const attributes = generateColumnAttributes(
            key,
            property,
            isRequired,
            isColumnNullable,
            notes,
        );

        if (property.$comment) text.push(stringifyComment(property.$comment));
        const attributesText = attributes.stringify(', ');
        const hasLineBreak = attributesText.includes('\n');

        text.push(`${snakeCase(key)} ${columnType} [${attributesText}${hasLineBreak ? '\n' : ']'}`);
        hasLineBreak && text.push(']');
        count++;
    }

    if (count === 0) {
        raise(`Table ${tableName} (${schemaId}) does not have any columns defined`);
    }

    return text;
}

function generateColumnAttributes(
    _key: string,
    property: ForgedPropertySchema,
    isRequired: boolean,
    isColumnNullable: boolean,
    notes: boolean,
): TextBuilder {
    const result = new TextBuilder();

    const isPK = property.dbColumn?.pk ?? false;

    if (isPK) {
        result.push('pk');
    }

    const defaultValue = property.default;
    const hasDefaultValue = defaultValue !== undefined;

    if (isColumnNullable && isRequired) {
        if (!hasDefaultValue) result.push('default: null');
        // result.push('null'); <- by default
    } else if (isColumnNullable || !isRequired) {
        result.push('null');
    } else {
        result.push('not null');
    }

    if (hasDefaultValue) {
        const v =
            typeof defaultValue === 'string' &&
            defaultValue[0] !== '`' &&
            defaultValue[defaultValue.length - 1] !== '`'
                ? `"${defaultValue}"`
                : defaultValue;
        result.push(`default: ${String(v)}`);
    }

    if (notes && property.description) {
        result.push(buildNote(property.description, 2).stringify());
    }
    return result;
}

function generateIndexes(
    tableName: string,
    schemaProperties: PRec<ForgedPropertySchema>,
    entityIndexes: DBMLEntityOptions['indexes'],
    includeNotes: boolean,
    schemaId: string,
): TextBuilder {
    const text = new TextBuilder();
    const idx: PRec<
        {
            columnType: string;
            columns: string[];
            fields: string[];
        } & DBMLIndexOptions
    > = {};
    const normalizedTableName = tableName.replace(/\./g, '_');
    const dropIndexes = new Set<string>();

    function processIndex(
        source: DBMLIndexOptionsDef,
        column: string,
        key?: string,
        rawKey?: string,
    ) {
        if (!key) key = column;
        if (!rawKey) rawKey = key;
        const list = asArray(source);

        for (const item of list) {
            let indexName;

            if (typeof item === 'string') {
                indexName = item;
            } else if (item === true) {
                indexName = `ix_${normalizedTableName}_${key}`;
            } else if (item != null && typeof item === 'object') {
                if (item.name) {
                    indexName = item.name;
                } else {
                    const k = list.length === 1 ? key : `${key}_${item.type ?? 'btree'}`;
                    indexName = `ix_${normalizedTableName}_${k}`;
                }
            }

            if (indexName) {
                if (item === false) {
                    dropIndexes.add(indexName);
                } else {
                    if (!schemaProperties[column]) raise('`Column ${column} not found in schema`');

                    const field = substr(rawKey, '.');
                    const index = (idx[indexName] ||= {
                        ...(typeof item === 'object' ? item : {}),
                        columnType: getDBType(schemaProperties[column], schemaId),
                        columns: [],
                        fields: [],
                    });
                    if (field) index.fields.push(field);
                    index.columns.push(column);
                }
            }
        }
    }

    if (entityIndexes) {
        for (const [key, def] of Object.entries(entityIndexes)) {
            const prefix = substr(key, 0, '.');
            const column = prefix ? prefix : key;
            processIndex(
                //
                def,
                column,
                prefix ? key.replace(/\./g, '_') : key,
                key,
            );
        }
    }

    for (const [column, property] of Object.entries(schemaProperties)) {
        if (property?.dbIndex !== undefined) {
            processIndex(property.dbIndex, column);
        }
    }

    for (const [indexName, index] of entries(idx)) {
        if (!index) continue;
        const { columns, unique, pk, comment, note } = index;

        if (columns.length !== new Set(columns).size) {
            console.warn(`Duplicate columns in index ${indexName} for table ${tableName}`);
        }

        const options = new TextBuilder(`name: "${indexName}"`);

        const type =
            index.type || index.columnType === 'jsonb' || index.columnType.endsWith('[]')
                ? 'gin'
                : undefined;
        if (type) options.push(`type: ${type}`);
        if (pk) options.push(`pk`);
        if (!pk && unique) options.push('unique');
        // todo?
        // if (includeNotes && fields.length) {
        //     options.push(`note: 'Fields: ${fields.join(', ')};'`);
        // }
        if (includeNotes && note) options.push(buildNote(note, 3).stringify());
        if (comment) text.push(`// ${comment}`);

        text.push(`(${columns.map(snakeCase).join(', ')}) [${options.stringify(', ')}]`);
    }

    return text;
}

function readDBEntityName({ dbEntity }: ForgedEntitySchema): string | undefined {
    if (dbEntity) {
        if (typeof dbEntity === 'string') return dbEntity;
        if (dbEntity.name) return dbEntity.name;
    }
    return undefined;
}

function readDBEntityIndexes({ dbEntity }: ForgedEntitySchema): DBMLEntityOptions['indexes'] {
    if (dbEntity != null && typeof dbEntity === 'object') {
        return dbEntity.indexes;
    }
    return undefined;
}

function stringifyComment(value: string | undefined): string[] | undefined {
    if (!value) return undefined;
    return value.split('\n').map((item) => `// ${item.trim()}`);
}

function buildNote(value: string | undefined, mlLevel = 1): TextBuilder {
    const result = new TextBuilder();
    if (value) {
        if (value.includes('\n')) {
            result.push(`note:`);
            result.push([`'''`, ...value.split('\n'), `'''`], mlLevel);
        } else {
            result.push(`note: "${value}"`);
        }
    }
    return result;
}

function isNullable(property: Schema): boolean {
    return (
        property.anyOf?.some((item) => typeof item !== 'boolean' && item?.type !== 'null') ?? false
    );
}

let DereferenceSchemaCache: SchemaDereferenceSharedCache | undefined;

function getDBType(property: ForgedPropertySchema, schemaId: string): string {
    if (property.dbColumn?.type) return property.dbColumn.type;

    let type = property.type as string | undefined;
    if (!type && property.$ref) {
        raise(`Type ${JSON.stringify(property)} (${schemaId}) was not dereferenced`);
    }
    if (type && Array.isArray(type)) {
        type = type.filter((item) => item !== 'null')[0] as string | undefined;
    }
    if (type && typeof type !== 'string') {
        raise(`Invalid SchemaTypeName value type (${typeof type}) (expected string)`);
    }

    const variants = property.anyOf || property.oneOf;

    if (variants) {
        const nonNullType = variants.find(
            (item) => typeof item !== 'boolean' && item?.type !== 'null',
        );
        if (nonNullType) {
            return typeof nonNullType === 'object'
                ? getDBType(nonNullType, schemaId)
                : nonNullType.toString();
        }
    }

    if (type === 'string') {
        if (property.format === 'date-time') {
            return 'timestamptz';
        }

        return property.format === 'uuid' ? 'uuid' : 'text';
    }

    if (type === 'object' || type === 'array') {
        return 'jsonb';
    }

    return type ?? 'jsonb';
}

function listProperties(schema: Schema): {
    type: 'object';
    properties: PRec<ForgedPropertySchema>;
    required: string[];
} {
    if (schema.properties) {
        return {
            type: 'object',
            properties: schema.properties as PRec<ForgedPropertySchema>,
            required: schema.required || [],
        };
    }

    if (schema.oneOf) {
        // todo oneOf
        raise(`Schema ${JSON.stringify(schema)} has oneOf, but not implemented yet`);
    }

    if (!schema.anyOf) {
        return {
            type: 'object',
            properties: {},
            required: [],
        };
    }

    const propertiesMap: PRec<ForgedPropertySchema> = {};
    const requiredSet = new Set<string>();
    const initialized = false;

    for (const item of schema.anyOf) {
        if (typeof item !== 'object') continue;
        if (item.type !== 'object') continue;
        if (typeof item.properties !== 'object') continue;

        if (!initialized) {
            for (const name of Object.keys(item.properties)) {
                propertiesMap[name] = item.properties[name] as ForgedPropertySchema;
            }
            for (const name of item.required || []) {
                requiredSet.add(name);
            }
        } else {
            if (!isEqualKeys(propertiesMap, item.properties)) {
                console.warn('Properties are not identical across all members of anyOf');
            }
            for (const name of item.required || []) {
                if (!requiredSet.has(name)) {
                    console.warn('Required fields are not identical across all members of anyOf');
                }
            }
        }
    }

    return {
        type: 'object',
        properties: propertiesMap,
        required: Array.from(requiredSet),
    };
}

const DefaultScopeName = 'Default';
const DefaultColumnsOrder = ['id', 'uid', 'slug', 'name'];

type TableSpec = {
    name: string;
    value: string;
};

// todo escape notes!
// todo escape comments!
