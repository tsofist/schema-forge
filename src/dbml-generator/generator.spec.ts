import { unlink } from 'node:fs/promises';
import { env } from 'node:process';
import { asBool } from '@tsofist/stem/lib/as-bool';
import { noop } from '@tsofist/stem/lib/noop';
import { SchemaDefinitionInfoForType, SchemaDefinitionInfoKind } from '../definition-info/types';
import { forgeSchema } from '../schema-generator/forge';
import { loadJSONSchema } from '../schema-registry/loader';
import { createSchemaForgeRegistry } from '../schema-registry/registry';
import { ForgeSchemaResult } from '../types';
import { generateDBMLSpec } from './generator';

const KEEP_ARTEFACTS = asBool(env['SFG_KEEP_ARTEFACTS'], false); // TODO!

describe('DBML Generator', () => {
    const outputSchemaFile = './dbml.c1.generated.schema.tmp.json';
    const outputSchemaMetadataFile = './dbml.c1.generated.definitions.tmp.json';
    const schemaId = 'test-dbml';

    let forgeSchemaResult: ForgeSchemaResult | undefined;
    let registry: ReturnType<typeof createSchemaForgeRegistry>;

    beforeAll(async () => {
        forgeSchemaResult = await forgeSchema({
            allowUseFallbackDescription: false,
            explicitPublic: true,
            schemaId,
            tsconfigFrom: './tsconfig.build-test.json',
            sourcesDirectoryPattern: 'test-sources/dbml/c1',
            sourcesFilesPattern: ['service.api.ts', '*.api.ts', 'types.ts'],
            outputSchemaFile,
            outputSchemaMetadataFile,
        });
        registry = createSchemaForgeRegistry();
        const schema = await loadJSONSchema([outputSchemaFile]);
        registry.addSchema(schema);
    }, 10_000);

    afterAll(async () => {
        if (!KEEP_ARTEFACTS) {
            await unlink(outputSchemaFile).catch(noop);
            await unlink(outputSchemaMetadataFile).catch(noop);
        }
    });

    it('generated schema should be valid', () => {
        expect(forgeSchemaResult).toBeTruthy();
        expect(forgeSchemaResult!.schema.$id).toStrictEqual(schemaId);
        expect(forgeSchemaResult!.generatedTemporaryFiles.length).toStrictEqual(1);
        expect(forgeSchemaResult!.refs.length).toStrictEqual(13);
    });

    it('basic cases', () => {
        const definitions = registry.listDefinitions((info, keywords) => {
            return (
                info.name.endsWith('Basic') &&
                info.kind === SchemaDefinitionInfoKind.Type &&
                keywords.has('dbEntity')
            );
        }) as SchemaDefinitionInfoForType[];

        const dbml = generateDBMLSpec(
            registry,
            [
                {
                    scopeName: 'TestScope',
                    comment: 'Test comment for TestScope',
                    schemaId,
                    definitions,
                },
            ],
            {
                meta: {
                    name: 'TestProject',
                    comment: 'Test comment!',
                    note: [
                        //
                        'Multi-line note',
                        '',
                        '> Powered by [SchemaForge]',
                    ].join('\n'),
                },
                includeNotes: true,
            },
        );

        expect(dbml).toMatchSnapshot();
    });

    it('table meta cases', () => {
        const definitions = registry.listDefinitions((info, keywords) => {
            return (
                info.name.endsWith('TableMeta') &&
                info.kind === SchemaDefinitionInfoKind.Type &&
                keywords.has('dbEntity')
            );
        }) as SchemaDefinitionInfoForType[];

        const dbml = generateDBMLSpec(
            //
            registry,
            [{ schemaId, definitions }],
            { includeNotes: true },
        );

        expect(dbml).toMatchSnapshot();
    });

    it('index meta cases', () => {
        const definitions = registry.listDefinitions((info, keywords) => {
            return (
                info.name.endsWith('IndexMeta') &&
                info.kind === SchemaDefinitionInfoKind.Type &&
                keywords.has('dbEntity')
            );
        }) as SchemaDefinitionInfoForType[];

        const dbml = generateDBMLSpec(
            //
            registry,
            [{ schemaId, definitions }],
            { includeNotes: true },
        );

        expect(dbml).toMatchSnapshot();
    });
});
