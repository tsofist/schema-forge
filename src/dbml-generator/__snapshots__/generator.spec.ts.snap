// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`DBML Generator basic cases 1`] = `
"// Test comment!
Project TestProject {
  database_type: 'PostgreSQL'

  note:
    '''
    Multi-line note

    > Powered by [SchemaForge]
    '''
}

// This is a comment
// Important: This is a multi-line comment
Table some_schema.some_entity_1 {
  number number [not null]
  number_array jsonb [not null]
  bigint number [not null]
  bigint_array jsonb [not null]
  bool boolean [not null]
  bool_array jsonb [not null]
  string text [not null]
  string_array jsonb [not null]
  date timestamptz [not null]
  date_array jsonb [not null]

  note:
    '''
      Description for Some Database Entity Type
      **Important**: This is a multi-line description for Some Database Entity Type
      This description may use **Markdown** syntax
    '''
}

// This is a comment
// Important: This is a multi-line comment
Table some_schema.some_entity_2 {
  ref uuid [pk, not null, note: "Some reference."]
  extra_ref uuid [pk, not null, note: "Some reference."]
  uuid uuid [not null]
  int integer [not null, note:
    '''
    Integer value.

    References:   - PostgreSQL: less than bigint and greater than integer (without NaN and +/- Infinity)
    '''
  ]
  int_array jsonb [not null]
  float number [not null, note: "Floating-point number."]
  float_array jsonb [not null]
  complex jsonb [not null, note: "Easy version of Record type"]
  complex_array jsonb [not null]

  indexes {
    (ref) [name: "ix_some_schema_some_entity_2_ref", unique]
    (extra_ref) [name: "ix_some_schema_some_entity_2_extraRef", unique]
  }

  note:
    '''
      Description for Some Database Entity Type
      **Important**: This is a multi-line description for Some Database Entity Type
      This description may use **Markdown** syntax
    '''
}

Table some_schema.some_entity_3 {
  int integer [not null, note:
    '''
    Integer compatibility type.

    References:   - PostgreSQL: integer
    '''
  ]
  bigint integer [not null, note:
    '''
    Big integer compatibility type.

    References:   - PostgreSQL: bigint
    '''
  ]
  smallint integer [not null, note:
    '''
    Small integer compatibility type.

    References:   - PostgreSQL: smallint
    '''
  ]
  float text [not null, note:
    '''
    Safe exact floating-point number.

    References:   - PostgreSQL: float8, double precision, numeric(30,18)
    '''
  ]
  money text [not null, note:
    '''
    String representation of a floating point number with up to 2 decimal places.

    References:   - PostgreSQL: usual numeric(12,2) or up to numeric(15,2)
    '''
  ]
}

Table some_schema.some_entity_5 {
  uid uuid [not null]
  zdt text [not null, note:
    '''
    ISO zulu date-time string (GMT time zone).

    Rules:
    * Date: YES
    * Time: YES
    * UTC offset: NO
    * Milliseconds: Optional
    '''
  ]
  ldt text [not null, note:
    '''
    ISO local date-time string.

    Rules:
    * Date: YES
    * Time: YES
    * UTC offset: Optional
    * Milliseconds: Optional
    '''
  ]
  ld text [not null, note:
    '''
    ISO local date string.

    Rules:  * Date: YES  * Time: NO  * UTC offset: NO  * Milliseconds: NO
    '''
  ]
  zd text [not null, note:
    '''
    ISO zulu date-time string (GMT time zone).

    Rules:
    * Date: YES
    * Time: YES
    * UTC offset: NO
    * Milliseconds: Optional
    '''
  ]
  pi integer [not null, note: "Positive integer value."]
  pf number [not null, note: "Positive floating-point number."]
  psf text [not null, note:
    '''
    Safe positive floating-point number.

    References:   - PostgreSQL: float8, double precision, numeric(30,18)
    '''
  ]
  ci integer [not null, note:
    '''
    Integer compatibility type.

    References:   - PostgreSQL: integer
    '''
  ]
  cnnbi integer [not null, note:
    '''
    Non-negative big integer compatibility type.

    References:   - PostgreSQL: bigint
    '''
  ]
  csni integer [not null, note:
    '''
    Negative small integer compatibility type.

    References:   - PostgreSQL: smallint
    '''
  ]
}

// Test comment for TestScope
TableGroup TestScope {
  some_schema.some_entity_1
  some_schema.some_entity_2
  some_schema.some_entity_3
  some_schema.some_entity_5
}
"
`;

exports[`DBML Generator index meta cases 1`] = `
"Project Scratch {
  database_type: 'PostgreSQL'
}

Table some_schema.ix_01 {
  index00 number [not null]
  indexed01 number [not null]
  indexed02 number [not null]

  indexes {
    (index00) [name: "ix_some_schema_ix_01_index00"]
    (indexed01) [name: "ix_some_schema_ix_01_indexed01", note: "This is a single-line note for Index"]
    (indexed02) [name: "ix_some_schema_ix_01_indexed02", note:
      '''
      This is a multiline-line note for Index
      Yep, this is true
      ''']
  }
}
"
`;

exports[`DBML Generator table meta cases 1`] = `
"Project Scratch {
  database_type: 'PostgreSQL'
}

// This is a comment
// Important: This is a ***multi-line comment***
Table some_schema.some_entity_10 {
  stub text [not null]
}

// This is a single-comment
Table some_schema.some_entity_11 {
  stub text [not null]
}

// This is a multi-comment
// This is a different kind of multi-line comment
Table some_schema.some_entity_12 {
  stub text [not null]
}

// V2
// This is a comment
// Important: This is a multi-line comment
//
// Outlined description
Table some_schema.some_entity_13 {
  stub text [not null]

  note:
    '''
    V2
    Description for Some Database Entity Type
    **Important**: This is a multi-line description for Some Database Entity Type
    This description may use **Markdown** syntax
    '''
}

Table some_schema.some_entity_6 {
  stub text [not null]

  note: "Single line description via '@description' tag"
}

Table some_schema.some_entity_7 {
  stub text [not null]

  note: "Inline description"
}

Table some_schema.some_entity_8 {
  stub text [not null]

  note:
    '''
    tag
    Description for Some Database Entity Type
    **Important**: This is a multi-line description for Some Database Entity Type
    This description may use **Markdown** syntax
    '''
}

Table some_schema.some_entity_9 {
  stub text [not null]

  note:
    '''
    Inline Multi-line description. Description for Some Database Entity Type
    **Important**: This is a multi-line description for Some Database Entity Type This description may use **Markdown** syntax
    '''
}
"
`;
